src/main/java/org/broadinstitute/consent/http/service/DatasetRegistrationService.java

```diff
@@ -14,48 +14,138 @@
     this.gcsService = gcsService;
     this.elasticSearchService = elasticSearchService;
+    this.studyDAO = studyDAO;
   }
 
+  public Study findStudyById(Integer studyId) {
+    Study study = studyDAO.findStudyById(studyId);
+    if (Objects.isNull(study)) {
+      throw new NotFoundException("Study with ID " + studyId + " is not found");
+    }
+    return study;
+  }
+
+  /**
+   * This method takes an instance of a dataset registration schema and updates the study and
+   * datasets from it.
+   *
+   * @param registration The DatasetRegistrationSchemaV1.yaml
+   * @param user         The User updating the study
+   * @param files        Map of files, where the key is the name of the field
+   * @return The updated Study
+   */
+  public Study updateDatasetsFromRegistration(
+      Integer studyId,
+      DatasetRegistrationSchemaV1 registration,
+      User user,
+      Map<String, FormDataBodyPart> files)
+      throws Exception {
+    Map<String, BlobId> uploadedFileCache = new HashMap<>();
+    List<FileStorageObject> uploadFiles = uploadFilesForStudy(files, uploadedFileCache, user);
+    List<DatasetServiceDAO.DatasetUpdate> datasetUpdates = new ArrayList<>();
+    List<DatasetServiceDAO.DatasetInsert> datasetInserts = new ArrayList<>();
+    // Dataset updates and inserts:
+    IntStream.range(0, registration.getConsentGroups().size())
+        .forEach(idx -> {
+          ConsentGroup cg = registration.getConsentGroups().get(idx);
+          if (Objects.nonNull(cg.getDatasetId())) {
+            Dataset existingDataset = datasetDAO.findDatasetById(cg.getDatasetId());
+            try {
+              DatasetUpdate datasetUpdate = new DatasetUpdate(
+                  cg.getConsentGroupName(),
+                  existingDataset.getNeedsApproval(),
+                  existingDataset.getActive(),
+                  existingDataset.getDacId(),
+                  convertConsentGroupToDatasetProperties(cg)
+              );
+              DatasetServiceDAO.DatasetUpdate update = createDatasetUpdate(
+                  cg.getDatasetId(),
+                  user,
+                  datasetUpdate,
+                  files,
+                  uploadedFileCache
+              );
+              datasetUpdates.add(update);
+            } catch (Exception e) {
+              logException(e);
+            }
+          } else {
+            try {
+              DatasetServiceDAO.DatasetInsert insert = createDatasetInsert(
+                  registration,
+                  user,
+                  files,
+                  uploadedFileCache,
+                  idx
+              );
+              datasetInserts.add(insert);
+            } catch (Exception e) {
+              logException(e);
+            }
+          }
+        });
+
+    List<StudyProperty> studyProps = convertRegistrationToStudyProperties(registration);
+    DatasetServiceDAO.StudyUpdate studyUpdate = new StudyUpdate(
+        registration.getStudyName(),
```

Comment:
I think there may be some discrepancy between the GET study endpoint and this PUT API. This endpoint is looking for fields on the study `studyName`, `consentGroups` and `studyDescription` where those fields are not the same on the `getStudyById` endpoint. I'm not sure if this problem is within the scope of this ticket but I am getting a bad request while testing with the GET. However, I also know that the data is corrupted in the studies so it may be something to do with that as well.

Reply:
The GET and the PUT are intentionally different. The GET should give you a study as defined by our Study class, but the PUT is expecting and instance of dataset registration schema which is very different from a Study